src/common/summarize-profile.js

```diff
@@ -118,37 +119,51 @@ function sampleCategorizer(thread) {
   const categorizeFuncName = functionNameCategorizer();
   const stackCategoryCache = new Map();
 
-  return function categorizeSampleStack(stackIndex) {
+  function computeCategory(stackIndex) {
     if (stackIndex === null) {
       return 'uncategorized';
     }
 
-    let category = stackCategoryCache.get(stackIndex);
-    if (category !== undefined) {
-      return category;
-    }
-
     const frameIndex = thread.stackTable.frame[stackIndex];
     const implIndex = thread.frameTable.implementation[frameIndex];
     if (implIndex !== null) {
-      // script.baseline or script.ion
-      category = 'script.' + thread.stringTable._array[implIndex];
-      stackCategoryCache.set(stackIndex, category);
-      return category;
+      // script.execute.baseline or script.execute.ion
+      return 'script.execute.' + thread.stringTable._array[implIndex];
     }
-    
+
     const funcIndex = thread.frameTable.func[frameIndex];
     const name = thread.stringTable._array[thread.funcTable.name[funcIndex]];
-    category = categorizeFuncName(name);
-    if (category !== false) {
-      stackCategoryCache.set(stackIndex, category);
+    const category = categorizeFuncName(name);
+    if (category !== false && category !== 'wait') {
+      return category;
+    }
+
+    const prefixCategory = categorizeSampleStack(thread.stackTable.prefix[stackIndex]);
+    if (category === 'wait') {
+      if (prefixCategory === 'uncategorized') {
+        return 'wait';
+      }
+      if (prefixCategory.endsWith('.wait') || prefixCategory === 'wait') {
+        return prefixCategory;
+      }
+      return prefixCategory + '.wait';
+    }
+
+    return prefixCategory;
+  }
+
```

Comment:
nit: It's quite nice how you are separating out the caching here, but since `stackCategoryCache` is only used with this function, perhaps it would be nice to locate its declaration right next to the function.